File: Days.32/C++/3.Infix-to-Prefix.cpp
// Infix ----> Prefix

#include<iostream>
#include<vector>
using namespace std;

int Precedence(char Operator)
{
    if (Operator=='^')
    {
        return 3;
    }
    else if(Operator=='*' or Operator=='/')
    {
        return 2;
    }
    else if(Operator=='+' or Operator=='-')
    {
        return 1;
    }
    else{
        return 0;
    }
}

string InfixtoPrefix(string &str)
{
    vector<string> Operand;
    vector<char> Operator;
    string a,b,c,temp;
    for(int i=0; i<str.size(); i++)
    {
        if (str[i]=='(')
        {
            Operator.push_back(str[i]);
        }
        else if(str[i]==')')
        {
            while (!Operator.empty() and Operator.back()!='(')
            {
                b=Operand.back();
                Operand.pop_back();
                a=Operand.back();
                Operand.pop_back();
                c=Operator.back()+a+b;
                Operand.push_back(c);
                Operator.pop_back();
            }
            Operator.pop_back();
        }
        else if(str[i]=='+' or str[i]=='-' or str[i]=='*' or str[i]=='/' or str[i]=='^')
        {
            while (!Operator.empty() and Operator.back()!='(' and Precedence(str[i])<=Precedence(Operator.back()))    
            {
                b=Operand.back();
                Operand.pop_back();
                a=Operand.back();
                Operand.pop_back();
                c=Operator.back()+a+b;
                Operand.push_back(c);
                Operator.pop_back();
            }
            Operator.push_back(str[i]);
        }
        else
        {
            temp=str[i];
            Operand.push_back(temp);
        }
    }
    while (!Operator.empty())
    {
        b=Operand.back();
        Operand.pop_back();
        a=Operand.back();
        Operand.pop_back();
        c=Operator.back()+a+b;
        Operand.push_back(c);
        Operator.pop_back();
    }
    return Operand.back();
}


int main()
{
    string str;
    cin>>str;
    string ans=InfixtoPrefix(str);
    cout<<ans<<"\n";
    return 0;
}

File: Days.32/Python/3.Infix-to-Prefix.py
# Infix ----> Prefix

def Precedence(operator):
    if operator=='^':
        return 3
    elif operator=='*' or operator=='/':
        return 2
    elif operator=='+' or operator=='-':
        return 1
    return 0


def InfixtoPrefix(str):
    operand=[]
    operator=[]
    for i in range(len(str)):
        if str[i]=='+' or str[i]=='-' or str[i]=='*' or str[i]=='/' or str[i]=='^':
            while operator and Precedence(operator[-1])>=Precedence(str[i]):
                b=operand.pop()
                a=operand.pop()
                operand.append(operator[-1]+a+b)
                operator.pop()
            operator.append(str[i])
        elif str[i]=='(':
            operator.append(str[i])
        elif str[i]==')':
            while operator[-1]!='(':
                b=operand.pop()
                a=operand.pop()
                operand.append(operator[-1]+a+b)
                operator.pop()
            operator.pop()
        else:
            operand.append(str[i])
    while operator:
        b=operand.pop()
        a=operand.pop()
        operand.append(operator[-1]+a+b)
        operator.pop()
    return operand[-1]

def main():
    str=input()
    ans=InfixtoPrefix(str)
    print(ans)


main()

